### PR TITLE
SPLAT-1434: aws: add support of BYO public ipv4 pool

### PR DESCRIPTION
**Pull request by mtulio**
Introduce support of AWS BYO IPv4 (Public IPv4 Pool) when creating clusters with publish strategy `External`.

The Public IPv4 Pool is an alternative for customers to mitigate the new AWS charges for Public IPv4 (Elastic IPs) ,provided by Amazon, when the customer has the custom IPv4 pools managed by AWS.

Refs:
- https://issues.redhat.com/browse/SPLAT-1434
- https://issues.redhat.com/browse/OCPSTRAT-1154
- https://github.com/openshift/release/pull/48467

### CODE REVIEW

File: docs/user/aws/custom_public_ipv4.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+- Public IPv4 Pool Provisioned in the Account
+- N Public Ips available in the pool
+> TODO steps to check the total ips available
+
+- Permissions?
+- Should the BYO VPC flow be impacted?
+
+## Steps
+
+- Create the install config settin the field `platform.aws.publicIpv4PoolId`, and create the cluster:
```
Comment (rvanderp3) [inline docs/user/aws/custom_public_ipv4.md:16]:
```suggestion
- Create the install config setting the field `platform.aws.publicIpv4PoolId`, and create the cluster:
```

File: data/data/aws/cluster/vpc/variables.tf [hunk 2]
```diff
@@ -59,4 +59,9 @@ variable "public_subnets" {
 variable "private_subnets" {
   type        = list(string)
   description = "Existing private subnets into which the cluster should be installed."
+}
+
+variable "public_ipv4_pool" {
+  type        = string
+  description = "An Public IPv4 Pool"
```
Comment (rvanderp3) [inline data/data/aws/cluster/vpc/variables.tf:66]:
```suggestion
  description = "A Public IPv4 Pool"
```

**Comment by mtulio** [comment]
> /test e2e-aws-ovn-public-ipv4-pool

First job run has been installed successfully, allocating the IP from the Public IP Pool.

/test e2e-aws-ovn-public-ipv4-pool

**Comment by patrickdillon** [comment]
This looks great. Do we need to do anything to release the IP address (on destroy, cluster or bootstrap)? Or does that happen automagically?

**Comment by mtulio** [comment]
> This looks great.
Thanks!

> Do we need to do anything to release the IP address (on destroy, cluster or bootstrap)? Or does that happen automagically?

No, Terraform does all the work. The Elastic IP (aws_eip) is a Terraform resource, we are just setting the attribute to consume the EIP from a different pool than Amazon provided, so once terraform deletes the EIP resource the Public IPv4 will be released from the BYO IP pool.

**Comment by mtulio** [comment]
> once terraform deletes the EIP resource the Public IPv4 will be released from the BYO IP pool.

Correct that statement, keeping the same idea of releasing: It's not Terraform that deletes the EIP in destroy flow, it is the [regular destroy flow by installer](https://github.com/openshift/installer/blob/master/pkg/destroy/aws/ec2helpers.go#L138-L139). Once EIP is deleted, the IP is returned to the custom pool.

**Comment by mtulio** [comment]
Hey PTAL for WG BYOIPv4?

/assign @patrickdillon @sadasu

File: pkg/asset/installconfig/aws/ec2.go [hunk 3]
```diff
@@ -28,3 +29,24 @@ func DescribeSecurityGroups(ctx context.Context, session *session.Session, secur
...
+	client := ec2.New(session, aws.NewConfig().WithRegion(region))
+
+	cctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	defer cancel()
+
+	poolOutputs, err := client.DescribePublicIpv4PoolsWithContext(cctx, &ec2.DescribePublicIpv4PoolsInput{PoolIds: []*string{aws.String(poolID)}})
+	if err != nil {
+		return nil, err
+	}
+	if len(poolOutputs.PublicIpv4Pools) == 0 {
```
Comment (barbacbd) [inline pkg/asset/installconfig/aws/ec2.go:44]:
```suggestion
	if len(poolOutputs.PublicIpv4Pools) != 1 {
	    return nil, fmt.Errorf("expected exactly 1 IPv4 Pool, found %d: %s", len(poolOutputs.PublicIpv4Pools), poolID)
```

File: data/data/aws/bootstrap/main.tf [hunk 4]
```diff
@@ -249,3 +249,11 @@ resource "aws_security_group_rule" "bootstrap_journald_gateway" {
   from_port   = 19531
   to_port     = 19531
 }
+
```
Comment (barbacbd) [inline data/data/aws/bootstrap/main.tf:252]: There are a lot of terraform changes. With the removal of terraform and addition of cluster api should we consider moving this over cluster api after the base installs work?

Comment (mtulio) [reply data/data/aws/bootstrap/main.tf:252]: @barbacbd yes, cluster API work is already in progress. The plan is to deliver both changes: https://issues.redhat.com/browse/OCPSTRAT-1154

**Review by sadasu — APPROVED** [review]

**Pull request merged**